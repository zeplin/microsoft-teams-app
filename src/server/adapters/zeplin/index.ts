export * from "./zeplin";
export * from "./types";
