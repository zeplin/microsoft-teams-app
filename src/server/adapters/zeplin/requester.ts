import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { ZeplinError } from "./ZeplinError";
import { INTERNAL_SERVER_ERROR } from "http-status-codes";

export class Requester {
    private instance: AxiosInstance;
    constructor(config: AxiosRequestConfig) {
        this.instance = axios.create(config);
    }

    async delete(url: string, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.delete(url, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async get<T = object>(url: string, config?: AxiosRequestConfig): Promise<T> {
        try {
            const { data } = await this.instance.get<T>(url, config);
            return data;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async patch(url: string, data: object, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.patch(url, data, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async post(url: string, data: object, config?: AxiosRequestConfig): Promise<string> {
        try {
            const { data: { id } } = await this.instance.post<{ id: string }>(url, data, config);
            return id;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }
}
