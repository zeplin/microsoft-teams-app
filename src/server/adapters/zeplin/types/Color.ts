export interface Color {
    id: string;
    created: number;
    name: string;
    r: number;
    g: number;
    b: number;
    a: number;
}
