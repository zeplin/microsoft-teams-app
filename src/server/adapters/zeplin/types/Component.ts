import { Thumbnails } from "./Thumbnails";

export interface Component {
    id: string;
    name: string;
    created: number;
    updated?: number;
    description?: string;
    image?: {
        width: number;
        height: number;
        original_url?: string;
        thumbnails?: Thumbnails;
    };
    section?: {
        id: string;
        group?: {
            id: string;
        };
    };
}
