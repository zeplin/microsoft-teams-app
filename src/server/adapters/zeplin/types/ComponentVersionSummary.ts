import { User } from "./User";
import { Thumbnails } from "./Thumbnails";

enum DestinationType {
    SCREEN = "screen",
    PREVIOUS = "previous",
}

export interface ComponentVersionSummary {
    id: string;
    created: number;
    creator?: User;
    commit?: {
        color: {
            r: number;
            g: number;
            b: number;
            a: number;
        };
        message?: string;
        author?: User;
    };
    image_url?: string;
    thumbnails?: Thumbnails;
    background_color?: {
        r: number;
        g: number;
        b: number;
        a: number;
    };
    source: string;
    width: number;
    height: number;
    density_scale: number;
    links: {
        rect: {
            x: number;
            y: number;
            width: number;
            height: number;
        };
        destination: {
            name?: string;
            type: DestinationType;
        };
    }[];
}
