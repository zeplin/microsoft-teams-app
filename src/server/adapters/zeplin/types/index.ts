export * from "./Organization";
export * from "./Project";
export * from "./Styleguide";
export * from "./RemPreferences";
export * from "./WebhookEvents";
export * from "./WebhookResourceType";
export * from "./ZeplinApplication";
export * from "./ProjectWebhook";
export * from "./StyleguideWebhook";
export * from "./User";
export * from "./WebhookStatus";
export * from "./WebhookUrlHealth";
