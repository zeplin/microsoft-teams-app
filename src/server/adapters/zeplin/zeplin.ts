import { ProjectWebhooks } from "./ProjectWebhooks";
import { Requester } from "./requester";
import { StyleguideWebhooks } from "./StyleguideWebhooks";

interface ZeplinOptions {
    url: string;
    webhookSecret: string;
}

class Zeplin {
    public projectWebhooks!: ProjectWebhooks;
    public styleguideWebhooks!: StyleguideWebhooks;

    init({ url, webhookSecret }: ZeplinOptions): void {
        const requester = new Requester({ baseURL: `${url}/v1/` });
        this.projectWebhooks = new ProjectWebhooks(requester, webhookSecret);
        this.styleguideWebhooks = new StyleguideWebhooks(requester, webhookSecret);
    }
}

export const zeplin = new Zeplin();
