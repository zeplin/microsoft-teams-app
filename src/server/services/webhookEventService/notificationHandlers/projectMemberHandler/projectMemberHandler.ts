import { NotificationHandler } from "../NotificationHandler";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { ProjectMemberInviteEvent, WebhookEvent } from "../../../../adapters/zeplin/types";
import { LONG_DELAY } from "../constants";
import { md } from "../md";

class ProjectMemberHandler extends NotificationHandler<ProjectMemberInviteEvent> {
    delay = LONG_DELAY;

    private getText(events: ProjectMemberInviteEvent[]): string {
        const [{
            payload: {
                context: {
                    project: {
                        name: projectName
                    }
                },
                resource: {
                    data: {
                        user: {
                            username
                        }
                    }
                }
            }
        }] = events;
        return events.length === 1
            ? md`**${username || "A new member"}** just joined _${projectName}_.`
            : md`**${events.length} new members** just joined _${projectName}_`;
    }

    getTeamsMessage(events: ProjectMemberInviteEvent[]): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: "Say hi ðŸ‘‹"
            }
        });
    }

    shouldHandleEvent(event: WebhookEvent): event is ProjectMemberInviteEvent {
        return event.payload.action === "invited";
    }
}

export const projectMemberHandler = new ProjectMemberHandler();
