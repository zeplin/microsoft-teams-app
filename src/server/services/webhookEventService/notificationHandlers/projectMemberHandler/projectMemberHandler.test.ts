import { projectMemberHandler } from "./projectMemberHandler";
import { ProjectMemberInviteEvent } from "../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    projectName?: string;
    userId?: string;
    userName?: string;
}

function getDummyEvent({
    projectName = "Project Microsoft Teams Integration",
    userId = "userId",
    userName = "dirtybit"
}: GetDummyEventParams = {}): ProjectMemberInviteEvent {
    return {
        payload: {
            action: "invited",
            context: {
                project: {
                    id: "projectId",
                    name: projectName
                }
            },
            resource: {
                data: {
                    user: {
                        id: userId,
                        username: userName
                    }
                }
            }
        }
    } as ProjectMemberInviteEvent;
}

describe("projectMemberHandler", () => {
    describe("getTeamsMessage", () => {
        it("should match snapshot when there is only one event", () => {
            expect(projectMemberHandler.getTeamsMessage([getDummyEvent()])).toMatchSnapshot();
        });

        it("should match snapshot when there are more than one events", () => {
            expect(projectMemberHandler.getTeamsMessage([
                getDummyEvent(),
                getDummyEvent({
                    userId: "userId2",
                    userName: "ergunsh"
                })
            ])).toMatchSnapshot();
        });
    });
});
