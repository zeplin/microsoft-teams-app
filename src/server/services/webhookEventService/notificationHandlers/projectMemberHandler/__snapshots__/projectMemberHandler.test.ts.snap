// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectMemberHandler getTeamsMessage should match snapshot when there are more than one events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 teammates** just joined _Project Microsoft Teams Integration_",
  "sections": [
    {
      "text": "Say hi ðŸ‘‹",
      "images": []
    }
  ]
}
`;

exports[`projectMemberHandler getTeamsMessage should match snapshot when there is only one event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**dirtybit** just joined _Project Microsoft Teams Integration_.",
  "sections": [
    {
      "text": "Say hi ðŸ‘‹",
      "images": []
    }
  ]
}
`;
